src/main/java/graphql/execution/Execution.java

```diff
@@ -77,8 +79,20 @@ public CompletableFuture<ExecutionResult> execute(Document document, GraphQLSche
         List<VariableDefinition> variableDefinitions = operationDefinition.getVariableDefinitions();
 
         CoercedVariables coercedVariables;
+        Supplier<NormalizedVariables> normalizedVariableValues;
         try {
-            coercedVariables = ValuesResolver.coerceVariableValues(graphQLSchema, variableDefinitions, inputVariables, executionInput.getGraphQLContext(), executionInput.getLocale());
+            coercedVariables = ValuesResolver.coerceVariableValues(graphQLSchema,
+                    variableDefinitions,
+                    inputVariables,
+                    executionInput.getGraphQLContext(),
+                    executionInput.getLocale());
+
+            normalizedVariableValues = FpKit.intraThreadMemoize(() ->
```

Comment:
I made this a supplier - if not one asks for normalised values of directives - then no work will be done.

QueryDirectivesIpl is also lazy by design -so this wont be asked for until some one asks for QueryDirectives

Reply:
Nice